### PR TITLE
Add static analysis

### PR DESCRIPTION
**Pull request by zanebeckwith**
Multiple static (and 'dynamic') analysis tools added here:
- ~~Valgrind sgcheck~~
- Clang scan-build
- cppcheck
- Google's address and undefined-behavior sanitizers
- Coverity

Also, all these new tools triggered a restructuring of the .travis setup, to use multiple jobs in order to make the logs (and any failing builds) easier to disentangle.

### CODE REVIEW

**Comment by zanebeckwith** [comment]
The `sgcheck` tool has failed to pick up all of the simple bugs I've thrown at it, at least when run on `travis-ci`. When I run a local version, it finds things. So, I'm afraid there's something wrong with the local installation of `valgrind` that I'm doing for `travis`. 

Further, the `sgcheck` job takes a long time to complete (due to installing `valgrind` locally). And, even when I get `sgcheck` to find something on a local machine, another one of the tools finds it too.

So, I'm wondering if we shouldn't include it at this time. Maybe include it later.

File: .travis.yml [hunk 1]
```diff
@@ -14,53 +14,108 @@
...
-  - sudo apt-get -qq update
-  - sudo apt-get install -y valgrind
+jobs:
+  include:
 
-  - git submodule update --init --recursive
-  - mkdir build
-  - cd build
+  # Release build, gcc
+  - if: not branch = coverity_scan
```
Comment (drbild) [inline .travis.yml]: Maybe let all these checks run on the `coverity_scan` branch too, just keep the `.travis.yml` more simple.  All the `if not`s distract from the common case.  It doesn't hurt anything to run all the checks - just a few a extra minutes of build time.

**Comment by drbild** [comment]
Other than the one comment, LGTM. 

It'd be nice to have the tools as build targets (e.g., `cd build; cmake ..; make cppcheck`) eventually, rather than needing to look up the command line incantation in the README or in the corresponding `.travis/*.sh` script.   E.g., a (possibly over-engineered) method for `cppcheck`: https://github.com/rpavlik/cmake-modules/blob/master/CppcheckTargets.cmake

File: .travis.yml (hunk 1 above)
Comment (zanebeckwith) [reply .travis.yml]: 👍

**Comment by zanebeckwith** [comment]
I agree on the build targets. I'm going to let it remain as-is, and see which ones are useful to us, and maybe in the future we can make those ones easier to run.

**Comment by zanebeckwith** [comment]
And, as much as it pains me (I have a soft spot for all of Valgrind), I'm going to remove sg-check from our tests, at least for now. It just gums up the build, and, like I said, my local installation must have been incorrect because it didn't seem to find anything.

**Pull request merged**